### PR TITLE
[WIP] Update core object

### PR DESCRIPTION
**Pull request by jgwhite**
**Current status:** still testing against addons.

### CODE REVIEW

File: lib/models/addon.js [hunk 1]
```diff
@@ -9,7 +9,6 @@ var SilentError = require('../errors/silent');
 var reexport    = require('../utilities/reexport');
 var escapeRegExp= require('../utilities/escape-regexp');
 
-
 var p                   = require('../preprocessors');
```
Comment (chadhietala) [inline lib/models/addon.js]: Can we rename this to preprocessors?

Comment (stefanpenner) [inline lib/models/addon.js]: :+1:

Comment (jgwhite) [inline lib/models/addon.js]: Totally.

**Comment by krisselden** [comment]
@jgwhite after working with the old core object, I'm a big fan of this.

**Comment by jgwhite** [comment]
This now passes but the changes are extensive. :grimacing: 

Making the switch without breaking our existing crop of addons involves two awkward contortions:

- [Project detects and uses the appropriate constructor signature for addons](https://github.com/jgwhite/ember-cli/commit/684c52169c3db80386e4a20f319d5f7f11cbd00f)
- [A modified version of core-object that makes the `_super` available as `_super.someMethod`](https://github.com/jgwhite/core-object/commit/029e3fbdf06159460f3f35951e17ffa1d39926e3)

Other concerns:

- Omitting the `this._super.apply(this, arguments)` in `init` will almost always break things
- The difference between `CoreObject#init` and `Ember.CoreObject#init` may cause confusion

**Comment by stefanpenner** [comment]
> Omitting the this._super.apply(this, arguments) in init will almost always break things

we can `fn.toString().indexOf('._super')`

**Comment by trabus** [comment]
Wow, awesome work! :clap:

**Comment by stefanpenner** [comment]
@jgwhite im nervous that this is getting way out of sync. Ping me, I would like to be sure to unblock you, or take the reigns before we diverge entirely, I would love to get this  in asap.

**Comment by jgwhite** [comment]
@stefanpenner I’ve been rebasing fairly regularly and I believe this is in a mergeable state right now.

**Comment by jgwhite** [comment]
Oops. Spoke too soon!

**Comment by stefanpenner** [comment]
@jgwhite we should sync up, ping me when you are free. Would love to unblock you :)

**Comment by stefanpenner** [comment]
@jgwhite looks like lots of progress was made again :) Care to share the current blockers?

**Comment by jgwhite** [comment]
@stefanpenner I’ve been trying to find a way to roll out the new CoreObject without changing exsting libraries but, the more I look at it, the more I think this will need a symlink-style unilateral upgrade. It’s mainly variations in expectations around `init` that cause trouble.

broccoli-caching-writer is a particularly off-beat culprit. What do you think of releasing a new-core-object-friendly version: https://github.com/ember-cli/broccoli-caching-writer/compare/ember-cli:master...jgwhite:upgrade-core-object

**Comment by stefanpenner** [comment]
@jgwhite im not quite sure what that external lib was causing grief. But it does not need to change (and also no longer uses core object itself)

**Comment by stefanpenner** [comment]
@jgwhite do you have interest in this, or should we close and move on?

**Comment by jgwhite** [comment]
@stefanpenner much as it pains me to say it — let’s close and revisit later.

**Comment by stefanpenner** [comment]
@jgwhite NP, future work :)

**Pull request closed without merging**